### PR TITLE
move copy of in-place ops to args/kwargs to the trace end

### PR DESCRIPTION
**Pull request by crcrpar**
## What does this PR do?

Fixes the comment of https://github.com/Lightning-AI/lightning-thunder/issues/657#issuecomment-2196237059.

- [ ] consider applying this to other cases such as `running_mean` and `running_var` of batch norm https://github.com/Lightning-AI/lightning-thunder/blob/8c5905fd1a93145e690791a7c7a3c3e10b16b32b/thunder/torch/__init__.py#L3419-L3453

For the function used in the added test case the functionalized trace and the extrace are 
```python
# Constructed by Functionalize in-place ops
import thunder
import thunder.core.prims as prims
import thunder.torch as ltorch
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a, b):
  # a: "cuda:0 f32[2, 2]"
  # b: "cuda:0 f32[2, 2]"
  t0 = ltorch.add(a, b, alpha=None)  # t0: "cuda:0 f32[2, 2]"
    # t0 = ltorch.add(a, b, alpha=None)  # t0: "cuda:0 f32[2, 2]"
      # t0 = prims.add(a, b)  # t0: "cuda:0 f32[2, 2]"

  # /home/mkozuki/ghq/github.com/Lightning-AI/lightning-thunder/snippet.py:7:       c = torch.exp(a)
  c = ltorch.exp(t0)  # c: "cuda:0 f32[2, 2]"
    # c = prims.exp(t0)  # c: "cuda:0 f32[2, 2]"

  # /home/mkozuki/ghq/github.com/Lightning-AI/lightning-thunder/snippet.py:8:       d = torch.tanh(b)
  d = ltorch.tanh(b)  # d: "cuda:0 f32[2, 2]"
    # d = prims.tanh(b)  # d: "cuda:0 f32[2, 2]"

  # /home/mkozuki/ghq/github.com/Lightning-AI/lightning-thunder/snippet.py:10:      e = c.view(-1)
  e = ltorch.view(c, -1)  # e: "cuda:0 f32[4]"
    # e = ltorch.reshape(c, (-1,))  # e: "cuda:0 f32[4]"
      # e = prims.reshape(c, (4,))  # e: "cuda:0 f32[4]"

  # /home/mkozuki/ghq/github.com/Lightning-AI/lightning-thunder/snippet.py:13:      e.add_(d.flatten())
  t5 = ltorch.flatten(d, 0, -1)  # t5: "cuda:0 f32[4]"
    # t5 = prims.reshape(d, (4,))  # t5: "cuda:0 f32[4]"
  t6 = ltorch.add(e, t5, alpha=None)  # t6: "cuda:0 f32[4]"
    # t6 = ltorch.add(e, t5, alpha=None)  # t6: "cuda:0 f32[4]"
      # t6 = prims.add(e, t5)  # t6: "cuda:0 f32[4]"
  t8 = ltorch.div(d, t0, rounding_mode=None, out=None)  # t8: "cuda:0 f32[2, 2]"
    # t8 = ltorch.div(d, t0, rounding_mode=None, out=None)  # t8: "cuda:0 f32[2, 2]"
      # t8 = ltorch.true_divide(d, t0)  # t8: "cuda:0 f32[2, 2]"
        # t8 = prims.div(d, t0)  # t8: "cuda:0 f32[2, 2]"

  # /home/mkozuki/ghq/github.com/Lightning-AI/lightning-thunder/snippet.py:16:      return c, d, e / 2.0
  result = ltorch.true_divide(t6, 2.0)  # result: "cuda:0 f32[4]"
    # result = prims.div(t6, 2.0)  # result: "cuda:0 f32[4]"
  t11 = prims.reshape(t6, (2, 2))  # t11: "cuda:0 f32[2, 2]"

  # /home/mkozuki/ghq/github.com/Lightning-AI/lightning-thunder/snippet.py:6:       a += b
  t1 = prims.copy_(t0, a)  # t1: "cuda:0 f32[2, 2]"

  # /home/mkozuki/ghq/github.com/Lightning-AI/lightning-thunder/snippet.py:16:      return c, d, e / 2.0
  return (t11, t8, result)

def computation(a, b):
  # a: "cuda:0 f32[2, 2]"
  # b: "cuda:0 f32[2, 2]"
  [result, t6, t8] = nvFusion0(a, b)
    # t0 = prims.add(a, b)  # t0: "cuda:0 f32[2, 2]"
    # c = prims.exp(t0)  # c: "cuda:0 f32[2, 2]"
    # d = prims.tanh(b)  # d: "cuda:0 f32[2, 2]"
    # e = prims.reshape(c, (4,))  # e: "cuda:0 f32[4]"
    # t5 = prims.reshape(d, (4,))  # t5: "cuda:0 f32[4]"
    # t6 = prims.add(e, t5)  # t6: "cuda:0 f32[4]"
    # t8 = prims.div(d, t0)  # t8: "cuda:0 f32[2, 2]"
    # result = prims.div(t6, 2.0)  # result: "cuda:0 f32[4]"
    # prims.copy_(t0, a)
  del a, b
  t11 = torch.reshape(t6, (2, 2))  # t11: "cuda:0 f32[2, 2]"
    # t11 = ltorch.reshape(t6, (2, 2))  # t11: "cuda:0 f32[2, 2]"
      # t11 = prims.reshape(t6, (2, 2))  # t11: "cuda:0 f32[2, 2]"
  del t6
  return (t11, t8, result)
```

### CODE REVIEW

**Review by lantiga — APPROVED** [review]
Would it make sense to try to move the copy as early as possible by analyzing usage, instead of defaulting to the end, since this will have memory implications?

Maybe we can open an issue for this. Also let's add an issue with the proposal for batchnorm.

**Comment by crcrpar** [comment]
> try to move the copy as early as possible by analyzing usage

just to check my understanding, would this mean it'd be better to have `t1 = prims.copy_(t0, a)` in the example above right after `t8 = ltorch.div(d, t0, rounding_mode=None, out=None)`?

That sounds nice

**Comment by lantiga** [comment]
That’s right, we might be able to del intermediates more aggressively if we do so.

**Review by t-vi — APPROVED** [review]
Looks good to me, thank you @crcrpar @lantiga 

Do we want to change the placement in this PR or merge and use the next?

**Comment by lantiga** [comment]
Let's merge first and follow up @t-vi
